### DISCUSSION TITLE
Nested submodules

### DISCUSSION THREAD

**Discussion by g-nardiello** [gocd/gocd#13038] [Q&A]
Dear all,

I am new to gocd and I am configuring my first project on my instance of gocd.

I have a project where a main git repository contains some git submodule repositories.

Each submodule could be a material and I can get some configurations by switching branch only to some of the submodules.

I tried to implement this configuration by creating a material for the main module and a material for each submodule, setting the right branch and the correct "alternate checkout path" for each repository, but it seems that gocd does not allow materials on nested folders on alternate checkout paths.

Example:
- repo1 (main repo) on branch main at "base" folder
- repo2 (submodule) on branch test at "base/sub2"
- repo3 (submodule) on branch dev at "base/sub3"

Could you please help me on how best configure my project to be automated with gocd?

Thank you

Best regards 
Giuseppe 

Edit: I have rewritten parts of the question to clarify its content, based on the feedback in your answers.

Note that the branch of each submodule repository could be changed in different configurations of the same project, ignoring the branch that is specified into the .gitmodules file. I think each configuration with a different combination of branches should be a different pipeline in gocd.

**Comment by chadlwilson** [comment]
Are you referring to git submodules? Or a.more generic notion of a module?

**Reply by g-nardiello** [reply]
Yes, I am referring to git submodules

**Reply by g-nardiello** [reply]
I have updated the question to explain it better.

**Reply by chadlwilson** [reply]
In my opinion this isn't really the way you are supposed to use git submodules within source controlled builds. The wrapping repo points to refs within the submodule at a particular point in time. In other words, the commit Sha of the wrapping repo implies specific commit shas for the submodules.

A GoCD material can point directly to the target repo of a submodule as a standalone material, or to the 'wrapping' git repo . If using the wrapping material it will checkout the submodules with the submodule ref versions that the 'wrapping' material implies.

So you cannot tell GoCD to ignore the branch or ref implied by the .gitmodules.

You could hack around with your own git scripting after clone/fetch of the wrapping material to update the submodule to whatever branch/ref you wanted, but you would lose the GoCD guarantees on integrity and GoCD would have no idea of the changes you'd made.

I would step back and ask why you are using submodules, and whether it would be better to model these as standalone repositories with standalone pipelines in GoCD that have gocd-modelled dependencies where each submodule produces an 'artifact' that is consumed by the 'wrapping' repository's build.

**Reply by g-nardiello** [reply]
Thank you for your message and your help. I will give you some additional information.

The original architecture provides for some components, which reside on different git repositories, to be located in subfolders of the main repository. These components update independently of the main repository, and by changing the branch of the submodule I have a different release of the project.

I am implementing the pipeline on GoCD as additional support and do not want to replace the previous architecture, so I want to avoid changing it at the moment.

The ideal solution I initially assumed was to disable the submodules from the main repository and checkout them via different materials where I specify the branch. However, if I understand correctly
- it is not possible to disable submodules
- it is not possible to use nested ‘alternate checkout paths’ between different materials

**Reply by chadlwilson** [reply]
> However, if I understand correctly
> 
>     * it is not possible to disable submodules

Correct. Just as submodules are deemed core functionality of git, and can't be disabled at `git config --global` level, GoCD also treats them as a core part of cloning the repo in the version controlled state.

> 
>     * it is not possible to use nested ‘alternate checkout paths’ between different materials

No, although when not using submodules, you could always do something like have three materials in your pipeline:

* repo1 parent material --> checkout to `<job working dir>/parent`
* repo2 submodule material --> checkout to `<job working dir>/repo2`
* repo3 submodule material --> checkout to `<job working dir>/repo3`

And then in the job tasks, have a first script task that does

```shell
rm -r parent/repo1 parent/repo2 # remove the submodule checkouts entirely
ln -s repo1 parent/repo1
ln -s repo2 parent/repo2
parent/run-the-build.sh
```

It's going a little against the spirit of submodules, and the changelog for the `parent` material would include submodule updates that you have overridden, but it'd broadly work OK.

**Reply by g-nardiello** [reply]
Ok, thank you for the clarification 😊

If I understand correctly, the solution that fits best is to manage the different branches within the single job. Could you confirm if I have understood correctly?

In that case, I could also change the branch of the submodule directly from the command line:
‘git submodule set-branch -b v1.0.0 xyz’
and then update the content with ‘git submodule sync’. In this case I could only use the submodule materials to monitor changes and as a trigger for the pipeline. Do you think this solution would work?

**Reply by chadlwilson** [reply]
In the "two submodules" example, my suggestion is three different materials cloned within and triggering a single pipeline and to ignore the submodule references entirely by overriding them - not to manually update/sync submodules.

When you have different combinations of branches you need to use, I would create a separate pipeline for each combination with the materials pointing to the relevant branches

If you do it the way you suggest above with a single material and some scripting to change submodules:
- a new GoCD build won't automatically trigger every time a given submodule repo's branch has new commits pushed to it
- GoCD won't show you the changes to the submodule repos anywhere, removing transparency.

Generally speaking GoCD is opinionated about continuous delivery and modelling of workflows. It's not **designed** to make switching branches easy even for normal materials - let alone submodules - as it comes from a philosophy of largely trunk-based development. if you are trying to do things outside its philosophy (hack with submodules, use branches to manage versions, override the way it clones materials) perhaps you don't need something like GoCD at all as you will be fighting how it is designed.

**Comment by sriramnrn** [comment]
On Wed, Aug 28, 2024 at 1:11 AM Giuseppe Nardiello ***@***.***>
wrote:

> Dear all,
>
> I am new to gocd and I am configuring my first project on my instance of
> gocd.
>
> I have a project where a main git repository contains some submodules.
>
> Each submodule could be a material and I can get some configurations by
> switching branch only to some of the submodules.
>
> I tried to implement this configuration by creating a material for the
> main module and a material for each submodule, setting the right branch for
> each repository, but it seems that gocd does not allow materials on nested
> folders.
>
Does the "alternate checkout path" setting meet your needs?

See: https://docs.gocd.org/current/configuration/admin_add_material.html

You could have the parent with an alternate checkout path set to "parent",
and then each sub module set to "parent/submodule1", "parent/submodule2",
and so on.

**Reply by g-nardiello** [reply]
I tried with an alternate checkout path but it seems that gocd does not support different materials with nested alternate checkout path, as the example you provided.

**Reply by g-nardiello** [reply]
I have updated the question to explain it better.
